File: dbm-ui/backend/flow/engine/bamboo/scene/mongodb/sub_task/backup.py
# -*- coding: utf-8 -*-
"""
TencentBlueKing is pleased to support the open source community by making 蓝鲸智云-DB管理系统(BlueKing-BK-DBM) available.
Copyright (C) 2017-2023 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at https://opensource.org/licenses/MIT
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
import logging.config
from typing import Dict, List, Optional, Tuple

from django.utils.translation import ugettext as _

from backend.flow.consts import MongoDBActuatorActionEnum
from backend.flow.engine.bamboo.scene.common.builder import SubBuilder
from backend.flow.engine.bamboo.scene.mongodb.sub_task.base_subtask import BaseSubTask
from backend.flow.plugins.components.collections.mongodb.exec_actuator_job2 import ExecJobComponent2
from backend.flow.utils.mongodb import mongodb_password
from backend.flow.utils.mongodb.mongodb_dataclass import ActKwargs, CommonContext
from backend.flow.utils.mongodb.mongodb_repo import MongoDBCluster, MongoDBNsFilter, MongoNodeWithLabel, ReplicaSet

logger = logging.getLogger("flow")


# BackupSubTask 处理某个Cluster的备份任务.
class BackupSubTask(BaseSubTask):
    """
    payload: 整体的ticket_data
    sub_payload: 这个子任务的ticket_data
    rs:
    backup_dir:
    """

    def __init__(self):
        pass

    @classmethod
    def make_kwargs(
        cls, payload: Dict, sub_payload: Dict, rs: ReplicaSet, cluster: MongoDBCluster, file_path: str
    ) -> dict:
        """备份 kwargs"""
        logger.debug("get_backup_node", sub_payload)
        nodes = rs.get_not_backup_nodes()
        if len(nodes) == 0:
            raise Exception("no backup node. rs:{}".format(rs.set_name))
        cls.parse_ns_filter(sub_payload)
        node = nodes[0]

        dba_user = "dba"
        dba_pwd = mongodb_password.MongoDBPassword().get_password_from_db(
            node.ip, int(node.port), node.bk_cloud_id, dba_user
        )["password"]

        ns_filter = sub_payload["ns_filter"]
        is_partial = MongoDBNsFilter.is_partial(ns_filter)
        oplog = payload["oplog"]
        if is_partial and oplog:
            raise Exception(_("oplog为True时, 不支持partial备份"))
        bk_dbm_instance = MongoNodeWithLabel.from_node(node, rs, cluster)
        sudo_account = ActKwargs().get_mongodb_os_conf()["user"]
        return {
            "set_trans_data_dataclass": CommonContext.__name__,
            "get_trans_data_ip_var": None,
            "bk_cloud_id": node.bk_cloud_id,
            "exec_ip": node.ip,
            "db_act_template": {
                "action": MongoDBActuatorActionEnum.Backup,
                "file_path": file_path,
                "exec_account": "root",  # 执行脚本的用户，一般是root
                "sudo_account": sudo_account,  # 执行actuator的用户，这里必须是必须用户.
                "payload": {
                    "bk_dbm_instance": bk_dbm_instance.__json__(),
                    "ip": node.ip,
                    "port": int(node.port),
                    "adminUsername": dba_user,
                    "adminPassword": dba_pwd,
                    "skipBackupSysDb": True,
                    "waitBackupSysTaskDone": True,
                    "backupType": "logical",
                    "args": {"backupNode": "", "isPartial": is_partial, "oplog": oplog, "nsFilter": ns_filter},
                },
            },
        }

    @classmethod
    def process_cluster(
        cls,
        root_id: str,
        ticket_data: Optional[Dict],
        sub_ticket_data: Optional[Dict],
        cluster: MongoDBCluster,
        file_path: str,
    ) -> Tuple[SubBuilder, List]:
        """
        backup a ReplicaSet or backup a ShardedCluster
        file_path 为 actuator_workdir 所在的分区MountPoint
        """

        # 创建子流程
        sb = SubBuilder(root_id=root_id, data=ticket_data)
        acts_list = []
        for rs in cluster.get_shards():
            kwargs = cls.make_kwargs(ticket_data, sub_ticket_data, rs, cluster, file_path)
            exec_ip = kwargs["exec_ip"]
            port = kwargs["db_act_template"]["payload"]["port"]
            acts_list.append(
                {
                    "act_name": _("MongoDB-备份-[{}:{}]".format(exec_ip, port)),
                    "act_component_code": ExecJobComponent2.code,
                    "kwargs": kwargs,
                }
            )

        sb.add_parallel_acts(acts_list=acts_list)
        # 将exec_ip取出来。传文件任务需要
        sub_bk_host_list = []
        for v in acts_list:
            sub_bk_host_list.append({"ip": v["kwargs"]["exec_ip"], "bk_cloud_id": v["kwargs"]["bk_cloud_id"]})

        return sb, sub_bk_host_list

File: dbm-services/common/bkdata-kafka-consumer/pkg/model/mysql_backup_result/backup.go
// TencentBlueKing is pleased to support the open source community by making 蓝鲸智云-DB管理系统(BlueKing-BK-DBM) available.
// Copyright (C) 2017-2023 THL A29 Limited, a Tencent company. All rights reserved.
// Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
// You may obtain a copy of the License at https://opensource.org/licenses/MIT
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.

package mysql_backup_result

import (
	"time"
)

type BackupMetaFileBase struct {
	// BackupId backup uuid 代表一次备份
	BackupId       string `json:"backup_id" db:"backup_id"`
	BackupType     string `json:"backup_type" db:"backup_type"`
	ClusterId      int    `json:"cluster_id" db:"cluster_id"`
	ClusterAddress string `json:"cluster_address" db:"cluster_address"`
	BackupHost     string `json:"backup_host" db:"backup_host"`
	BackupPort     int    `json:"backup_port" db:"backup_port"`
	MysqlRole      string `json:"mysql_role" db:"mysql_role"`
	// ShardValue 分片 id，仅 spider 有用
	ShardValue int    `json:"shard_value" db:"shard_value"`
	BillId     string `json:"bill_id" db:"bill_id"`
	// BkBizId 被清洗过了
	BkBizId         int    `json:"bk_biz_id" db:"bk_biz_id"`
	MysqlVersion    string `json:"mysql_version" db:"mysql_version"`
	DataSchemaGrant string `json:"data_schema_grant" db:"data_schema_grant"`
	// IsFullBackup 是否包含数据的全备
	IsFullBackup bool `json:"is_full_backup" db:"is_full_backup"`
	// BackupConsistentTime 备份的一致性时间点，逻辑备份是备份开始时间，物理备份是备份结束时间， format time.RFC3339
	BackupConsistentTime time.Time `json:"backup_consistent_time" db:"backup_consistent_time"`
	// BackupBeginTime use time.RFC3339
	BackupBeginTime time.Time `json:"backup_begin_time" db:"backup_begin_time"`
	BackupEndTime   time.Time `json:"backup_end_time" db:"backup_end_time"`

	// ConsistentBackupTime todo 为了字段兼容性，可以删掉
	ConsistentBackupTime time.Time `json:"consistent_backup_time" db:"consistent_backup_time"`
}

type ExtraFields struct {
	BkCloudId        int    `json:"bk_cloud_id" db:"bk_cloud_id"`
	FileRetentionTag string `json:"file_retention_tag" db:"file_retention_tag"`
	TotalFilesize    uint64 `json:"total_filesize" db:"total_filesize"`
	// TotalSizeKBUncompress 压缩前大小，如果是zstd压缩会提供压缩前大小，-1,0 都是无效值。这不是精确大小，可能存在四舍五入
	TotalSizeKBUncompress int64 `json:"total_size_kb_uncompress" db:"total_size_kb_uncompress"`
	EncryptEnable         bool  `json:"encrypt_enable" db:"encrypt_enable"`
	// StorageEngine 物理备份使用
	StorageEngine string `json:"storage_engine" db:"storage_engine"`
	TimeZone      string `json:"time_zone" db:"time_zone"`
	// BackupCharset 逻辑备份使用
	BackupCharset  string `json:"backup_charset" db:"backup_charset"`
	SqlMode        string `json:"sql_mode" db:"sql_mode"`
	BinlogFormat   string `json:"binlog_format" db:"binlog_format"`
	BinlogRowImage string `json:"binlog_row_image" db:"binlog_row_image"`
	// BackupTool command name xtrabackup / mydumper / mysqldump
	BackupTool string `json:"backup_tool" db:"backup_tool"`
}

type BinlogStatusInfo struct {
	// ShowMasterStatus 当前实例 show master status 输出，本机位点
	ShowMasterStatus *StatusInfo `json:"show_master_status"`
	// ShowSlaveStatus 显示的是当前实例的 master 的位点
	ShowSlaveStatus *StatusInfo `json:"show_slave_status"`
}

type StatusInfo struct {
	BinlogFile string `json:"binlog_file"`
	BinlogPos  string `json:"binlog_pos"`
	Gtid       string `json:"gtid"`
	MasterHost string `json:"master_host"`
	MasterPort int    `json:"master_port"`
}

type TarFileItem struct {
	FileName      string   `json:"file_name"`
	FileSize      int64    `json:"file_size"`
	FileType      string   `json:"file_type" enums:"schema,data,metadata,priv"`
	ContainFiles  []string `json:"contain_files"`
	ContainTables []string `json:"contain_tables"`
	// TaskId backup task_id
	TaskId string `json:"task_id"`
}

type IndexContent struct {
	BackupMetaFileBase
	// ExtraFields 这里不能展开
	ExtraFields

	// BinlogInfo show slave status / show master status
	BinlogInfo BinlogStatusInfo `json:"binlog_info" db:"binlog_info"`

	FileList []*TarFileItem `json:"file_list" db:"file_list"`
}
